Add Yaml to Azure Pipelines

pablonete opened the pull request.

hross committed:
Add azure pipelines CI.

Commit:
Publish test results.

Commit:
Make diff more significant (125+50 ~= 200)

Commit:
No need to setTimeout, let's `done` on the callback.

Commit:
Make diff more significant to avoid race conditions.

Commit:
Wait a bit more so default 5 retries did happen.

Commit:
Don't resume twice.
Check queue length on callback for safety, setTimeout may be delayed.

Commit:
Avoid time dependency, act on events instead.

Commit:
Make delays shorter to avoid "Timeout of 250ms exceeded"

Commit:
Shorten delay to avoid 200ms test timeout in browser tests.

Commit:
Shorten timeout to mitigate 200ms test timeout.

Commit:
Avoid dependency on timing to final assertion.

Commit:
Rewrite test to avoid setTimeout

Commit:
Add more diff on delays to enforce test behavior.

willsmythe commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,61 @@
+jobs:
```
I would name this file `.azure-pipelines.yml` so it sits next to the "steps" file at the top of the source tree

Commit:
Enable coveralls

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,52 @@
+    versionSpec: $(node_version)
+  displayName: 'Install Node.js'
+
+- task: Npm@1
```
Since they have a package-lock.json, you should call "npm ci" instead of "npm install" (it's faster and considered a best practice). The challenge is that "npm ci" isn't available in older versions of Node. So two ideas on how to do this:
1. Always upgrade npm to the latest version (regardless of the Node version). Note: I would make sure this is considered safe, i.e. the latest version of NPM works with Node 6.
2. Use a simple inline [bash] script to check the version of npm and run "npm ci" if the version is > 5.7.1 (or whatever version is considered the minimum for "npm ci"), otherwise run "npm install". 
   * Alternatively condition the step on your node_version variable, meaning you would have define two steps, but only one would run.

pablonete replied:
npm is always 6.5.0, so `ci` works for my entire node matrix.

pablonete replied:
Ha, it was true for Windows only, in Linux and OSX Node 6 brings with npm 3.3.10. So I'm updating it conditionally.

Commit:
Use Safari  for OSX browser tests.

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,52 @@
+  displayName: 'Install dependencies'
+
+- bash: |
+    node_modules/.bin/mocha test --reporter mocha-junit-reporter --reporter-options mochaFile=./test-results.xml
```
This feels brittle. I would use "npm run xxx" to call the appropriate scripts defined in package.json (passing the reporter, etc as additional args). Alternatively, add CI-specific scripts to package.json, like "mocha-test-ci" (or "mocha-node-test-ci" and "mocha-browser-test-ci") which then call the appropriate scripts (but with the additional arguments).

pablonete replied:
Correct. I just needed a double -- escape to let the arguments get in.

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,52 @@
+      make -j 4
+      npm run mocha-browser-test
+  displayName: 'Run browser tests'
+  condition: |
```
One problem with having a template shared by lots of different jobs is that you end up with complex condition statements when you need to limit steps to certain jobs. One way to reduce the complexity of the condition (i.e. make the YAML easier to understand and easier to extend later) is to have a variable control whether to do the step(s) or not. This variable would get set in the matrix(s) where it's needed. The condition on the step then becomes a lot simpler / more logical, for example: `condition: variables.test_browsers` (or whatever)

pablonete commented:
coveralls is publishing here: https://coveralls.io/repos/248778/builds

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,52 @@
+  condition: always()
+
+- bash: |
+      export DISPLAY=:99.0
```
Since the DISPLAY variable is only relevant on Linux, I would elevate it to a variable on the Linux job. I would then use it when you call Xvfb: 

```
/usr/bin/Xvfb ${DISPLAY} -ac >> /tmp/xvfb.out 2>&1 &
```

This avoids a problem down the road when someone changes the value of DISPLAY, but forgets to change the Xvfb call.

willsmythe commented on karma.conf.js:
```diff
@@ -7,12 +7,16 @@ module.exports = function(config) {
         },
-        reporters: ['mocha'],
+        reporters: ['mocha', 'junit'],
+        junitReporter: {
```
Should the junit reporter only be used during CI? Mainly asking since some people may not be OK with this file getting created when these tests are run locally. Alternatively, add `*test-results.xml` to `.gitignore`.

Commit:
Adding missing lib files to karma (only Safari complained).

Commit:
Rename with dot so it gets better placed in order.

Commit:
Use ci instead of install, so we stick with package-lock.json

Commit:
Use npm test instead of mocha directly. It just needed a double -- escape.

Commit:
Move DISPLAY to Linux only, with value set just once.

Commit:
Use variables for DISPLAY, browser tests and coveralls.

Commit:
Simplify steps with a browser name variable.

Commit:
Run junit reporter on CI only.

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,50 @@
+  displayName: 'Install dependencies'
+
+- bash: |
+    npm test -- -- --reporter mocha-junit-reporter --reporter-options mochaFile=./test-results.xml
```
Yah ... I haven't seen the double "--" before. For future reference, what requires this?

pablonete replied:
It's because we're calling a npm script that calls another npm script. Our params need to pass through both calls. I see its fragility, but I still prefer defining the xml file in the same place where it's consumed (the yml file). Let me know if you think otherwise, alternative would be to duplicate more on the package.json (w/wo ci reporter).

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,50 @@
+  displayName: 'Update npm'
+  condition: eq(variables['node_version'], '6.x')
+
+- task: Npm@1
```
This is probably more personal preference, but I avoid the "npm" task because it doesn't add much value, requires more lines (i.e. increases complexity), and you usually then end up with a mix of script and NPM task steps invoking NPM in different ways anyway. Like you have done, I like to have separate tasks for each NPM command (improves readability of the status/logs), but will just use "script" for each, for example:

```yaml
- script: npm ci
  displayName: 'Install dependencies'
```

pablonete replied:
I could avoid it here `npm ci`, but not the `npm i -g npm` that we need to get the `ci` command available. When I do it straightforwardly on a script step, it doesn't update the toolcache version. I tried different combinations, and had to maintain it finally through `Npm@1`

willsmythe commented on .azure-pipelines-steps.yml:
```diff
@@ -0,0 +1,50 @@
+  condition: always()
+
+- bash: |
+    export DISPLAY=$(DISPLAY)
```
`DISPLAY` should be an environment variable, meaning you don't need this EXPORT statement, right? If for some reason it's not already an environment variable (I am assuming matrix variables become environment variables), I would set it via the "env" property on the bash task.

pablonete replied:
I didn't know this, thanks.

willsmythe commented on .azure-pipelines.yml:
```diff
@@ -0,0 +1,72 @@
+jobs:
+
+- job: Windows
+  pool: 'Hosted VS2017'
```
Use pool/vmImage so you're more consistent with the Linux and macOS jobs:

```
pool:
    vmImage: 'vs2017-win2016'
```

willsmythe commented on test/cargoQueue.js:
```diff
@@ -46,7 +46,7 @@ describe('cargoQueue', () => {
                 call_order.push('callback ' + 3);
             });
-        }, 15);
+        }, 20);
```
So even if we don't introduce a test variable that acts as a multiplier for all/most timeout values, I would still increase these hard-coded values in a consistent way so it's easier to explain in the PR and doesn't feel like us arbitrarily picking values that work in our CI environments. For example, you multiplied each value in `delays` by 2 (which sounds reasonable), but then only increased this one by 1.3 (15 * 1.3), and the next by 2.67 (30 * 2.67).

I think it will be easier for them to accept these changes if the changes feel more uniform versus us finding values that happen to work in our CI system. Does that make sense?

pablonete replied:
I see your point. Change here was intended to make timing diff more significant, so one step happens earlier and the next happens later, mitigating race conditions.
Anyway, I'm rethinking all my timing changes.

pablonete replied:
When I try more systematic changes on timings, I get worse results. See this run:
https://dev.azure.com/pablo-nunez/Runner/_build/results?buildId=215&view=ms.vss-test-web.build-test-results-tab
Doubling delays doesn't fix `queue basics` (2 failures in OSX). This is because nothing enforces the order of execution of processes and callbacks, so we're subject to race conditions if delays are not significant enough. Their original delays are a bit random, and on my proposal I'm exaggerating them to enforce relative order. See better results on this branch:
https://dev.azure.com/pablo-nunez/Runner/_build/results?buildId=214&view=ms.vss-test-web.build-test-results-tab
So what's pending:
- Identify other tests that should be tweaked up.
- Investigate timeouts in Windows 10 browser tests.
- Fix autoInject for Safari in production code.

willsmythe commented on test/concat.js:
```diff
@@ -3,8 +3,6 @@ var {expect} = require('chai');
 var assert = require('assert');
 
 describe('concat', function() {
-    this.timeout(250);
```
What was the purpose of `this.timeout(250)` before? Was it setting a timeout for this particular test (i.e. fail the test if it doesn't run in under 250 milliseconds)?

pablonete replied:
It's part of my change to set 2000ms as a global timeout. Initially this line increased the timeout from default 200ms, but after my change, its meaning was the opposed. Still double checking the actual value of the 2000ms change.

willsmythe commented on test/es2017/asyncFunctions.js:
```diff
@@ -304,7 +304,7 @@ module.exports = function () {
     it('should handle async functions in autoInject', (done) => {
         async.autoInject({
             z: async function(){ return 0},
```
nit: add a space after `function` to match the style elsewhere (yah, I know you didn't make this change, but it sticks out after you made the change to remove "a " in the next line)

willsmythe commented on test/queue.js:
```diff
@@ -4,13 +4,10 @@ var assert = require('assert');
 
         var call_order = [];
-        var delays = [40,10,60,10];
+        var delays = [60,10,100,10];
```
Similar to my earlier comment, can you increase each value more consistently -- i.e. same multiplier. So for example: `80,20,120,20`

Commit:
DISPLAY should be already an environment variable.

Commit:
Don't use verbose npm@1 task.

Commit:
Use vmImage for consistency with other OS.

Commit:
Fix formatting.

Commit:
Make delays more significant to avoid race conditions.

Commit:
Stop using setTimeout and rely on events to push new items, so order is enforced.

Commit:
Fix async function serialization in Safari.

Commit:
Add more delay so 4 never completes before 3.

Commit:
Ensure 3 never completes before 4.

Commit:
Use drain instead of task count.
Trying to ensure done is called from Windows browser tests that fail frequently.

Commit:
Ensure 2 starts before 3 arrives.Avoid expected 'process 2 3' to equal 'process 2'

Commit:
Add retries to browser tests.

Commit:
Don't fail all on Mac & Windows browser tests.

Commit:
Ensure 3 happens before 2, it fails sparely on OSX.
AssertionError: expected [ 1, 2, 3 ] to deeply equal [ 1, 3, 2 ]

Commit:
Avoid unnecessary multilines.

Commit:
Use Edge for Windows browser tests.
